### ISSUE TITLE
docker ps cannot return, however docker ps -n num can success with containers info

### ISSUE THREAD

**Issue by GolfenGuo** [docker/docker#18279]
I'm working on a docker server which contains 98 containers.  After the docker daemon running several days, the command (docker ps) will hold forever until I Ctrl-C cancel it.  More interesting is, under this condition, I can get expected results by docker ps -n <NUM> .  The exactly NUM just likes a threshold, any number less than the NUM also works.  For example, if the NUM is 88, both docker ps -n 80 and docker ps -n 78 will work, but docker ps -n 89 will fail, likes below.

/home/ubuntu# docker ps -n=88 |wc -l
89

/home/ubuntu# docker ps -n=89 |wc -l

Also, there are several socket with CLOSE_WAIT state:

/home/ubuntu# netstat -p |grep 10.10.5.152
tcp6       1      0 10-10-138-100:5656      10.10.5.152:43917       CLOSE_WAIT  12598/docker
tcp6       1      0 10-10-138-100:5656      10.10.5.152:45837       CLOSE_WAIT  12598/docker
tcp6       0      0 10-10-138-100:5656      10.10.5.152:41800       ESTABLISHED 12598/docker
tcp6       0      0 10-10-138-100:5656      10.10.5.152:41799       ESTABLISHED 12598/docker

I'm suspect that the above issue relates the docker exec.  In my usage, we often run docker exec from a remote Server-B by binding to the docker IP:port endpoint of Server-A.  For a while, the Server-A will show the above issue.  For another Server-C which don't receive any docker exec work very well.

Can any one help me on this case.  I'm glad to assist you on any additional information if necessary.

The environment likes below:

/home/ubuntu# docker version
Client:
 Version:      1.8.1
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   d12ea79
 Built:        Thu Aug 13 02:35:49 UTC 2015
 OS/Arch:      linux/amd64

/home/ubuntu# docker info
Containers: 98
Images: 8283
Storage Driver: devicemapper
 Pool Name: docker-253:16-4194310-pool
 Pool Blocksize: 65.54 kB
 Backing Filesystem: extfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 174.8 GB
 Data Space Total: 536.9 GB
 Data Space Available: 352.4 GB
 Metadata Space Used: 325.2 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 1.822 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Data loop file: /data/var/vcap/data/docker/devicemapper/devicemapper/data
 Metadata loop file: /data/var/vcap/data/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.77 (2012-10-15)
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.13.0-68-generic
Operating System: Ubuntu 14.04 LTS
CPUs: 2
Total Memory: 11.74 GiB
Name: 10-10-138-100
ID: IC2A:VDXC:C6KB:OXDP:SOOH:SWU5:V6EC:3NZG:JWNI:C3FF:T5L2:M6DG

/home/ubuntu# uname -a
Linux 10-10-138-100 3.13.0-68-generic #111-Ubuntu SMP Fri Nov 6 18:17:06 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

**Comment by thaJeztah** [comment]
Are you still able to reproduce this on the current release? (docker 1.9.1)

**Comment by GolfenGuo** [comment]
There are 90+ containers running on the server, the upgrade have to involve containers downtime or migration that's risky to me.  Is it a already-known bug? or does the new version fix this issue?

**Comment by thaJeztah** [comment]
Some improvements/changes were made to performance of `docker ps`, but I cannot predict if those resolve the issue in your use case (performance issues can be very specific to an install and the way docker is used).

I can understand that taking down / restarting a production server for testing is not something you want to  do, but if you're able to test on a non-production server, that'd be great


/cc @calavera because I think he was looking into additional improvements in docker ps

**Comment by duglin** [comment]
@GolfenGuo I've tried a couple of times to reproduce this but haven't been able to. Is there any additional info you can provide?  Have you tried to reproduce it on another (clean) host?

**Comment by GolfenGuo** [comment]
@thaJeztah many thanks for your suggestion.  I can handle with the slow PERFORMANCE, however the docker ps will never get back is the key issue I was taking about.  Since the command (docker ps -n NUM) can get back very quick, I'm doubt that it may caused by some container data is broken, as a result, the docker ps pending on a broken data.

**Comment by GolfenGuo** [comment]
@duglin I'm running a container platform that contains many docker servers.  The most of severs have the same problem described above, and the common of them is that they are running many containers (around 70+).  Additionally, I think that the issue relates to (docker exec) since the problem don't shown on those servers that disabled the (docker exec).  

In my context, we invoke the (docker exec) from a remote client through binding to docker deamon IP:port.  And, the client also runs in a container that be destroyed every 15 mins.

**Comment by allencloud** [comment]
＋1

**Comment by duglin** [comment]
I've tried to reproduce this (yes using exec in there too) but I can't. Perhaps some of the recent changes that went in have fixed it. Please let us know if using the latest code still shows the issue or not. But, since there doesn't seem to be a way to reproduce this outside of your environment I'm not sure how to make progress on this. Perhaps you could reproduce it and produce a dump of all the threads so we can see where the 'ps' thread is at - that might give us a clue.

In the meantime, w/o something more concrete to go on I'm going to close this - you can reopen if there's more info that can help us track this one down.

**Comment by grogancolin** [comment]
I've seen this happen again:

`
ubuntu@jenkins-slave:~$ docker --version
Docker version 1.11.2, build b9f10c9
ubuntu@jenkins-slave:~$ uname -a
Linux jenkins-slave 3.13.0-91-generic #138~precise1-Ubuntu SMP Fri Jun 24 18:50:48 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
ubuntu@jenkins-slave:~$ docker ps -n 5
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS                     PORTS               NAMES
6f6f401c6594        ubuntu              "/bin/bash"         2 minutes ago       Exited (0) 2 minutes ago                       small_kowalevski
fcfe83d586a9        ubuntu              "-d"                2 minutes ago       Created                                        mad_cori
ubuntu@jenkins-slave:~$ docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
ubuntu@jenkins-slave:~$
`

If you need any more info, let me know, I'll try getting it to you.

**Comment by jagadeeswar725863** [comment]
Hi All,

we have almost 50 containers are running on my docker server. and we are using docker version 1.12.1. we are not even able to run below command.

sudo docker ps -n=3 |wc -l

========
[infraansi@dc1mznode1 ~]$ docker --version
Docker version 1.12.1, build 23cf638

[infraansi@dc1mznode1 ~]$ sudo docker info
Containers: 45
 Running: 40
 Paused: 0
 Stopped: 5
Images: 85
Server Version: 1.12.1
Storage Driver: devicemapper
 Pool Name: docker_storage-docker--pool
 Pool Blocksize: 524.3 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file:
 Metadata file:
 Data Space Used: 61.26 GB
 Data Space Total: 128.7 GB
 Data Space Available: 67.45 GB
 Metadata Space Used: 29.33 MB
 Metadata Space Total: 323 MB
 Metadata Space Available: 293.6 MB
 Thin Pool Minimum Free Space: 12.87 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: true
 Deferred Deleted Device Count: 1
 Library Version: 1.02.135-RHEL7 (2016-09-28)
Logging Driver: journald
Cgroup Driver: systemd
Plugins:
 Volume: local
 Network: host null bridge overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp selinux
Kernel Version: 3.10.0-514.10.2.el7.x86_64
Operating System: Red Hat Enterprise Linux
OSType: linux
Architecture: x86_64
CPUs: 72
Total Memory: 503.6 GiB
Name: dc1mznode1.gst.gov.in
ID: ECHZ:JLEK:25YE:TUJD:S4LW:GBNO:BBYQ:AYF4:JH67:FAZ5:AHFC:QKFG
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Insecure Registries:
 gstnregistry.dc1.internal.gst.gov.in
 127.0.0.0/8
[infraansi@dc1mznode1 ~]$

Kindly help some one on this.
Thanks in Advance.

**Comment by thaJeztah** [comment]
@jagadeeswar725863 docker 1.12 has reached EOL; also be aware that 1.12.1 has known vulnerabilities (1.12.6 is the latest patch release for 1.12.x).

I recommend updating to a supported version (17.09 or above); if you're still having this issue on a current release, open a new issue in the https://github.com/docker/for-linux issue tracker

**Comment by JeremyBesson** [comment]
Hi all, 

We face the same error in docker 1.12.6 with approximately 170 running containers.
To fix this we restart docker service and the command docker ps work again.

Is there someone who can confirm an upgrade to 17.09 fix this issue ? 

Thanks
